packages/mobile/src/firebase/firebase.ts

```diff
@@ -76,6 +77,58 @@ export const firebaseSignOut = async (app: ReactNativeFirebase.FirebaseApp) => {
   await app.auth().signOut()
 }
 
+export function* setupMessaging(action: AppMounted | AppUnmounted) {
+  if (!FIREBASE_ENABLED) {
+    // TODO(erdal) enable this once we have gcloud setup on CI
+    return
+  }
+
+  // Listen for notification messages while the app is open
+  const channelOnNotification: EventChannel<NotificationChannelEvent> = eventChannel((emitter) => {
+    const unsubscribe = () => {
+      Logger.info(TAG, 'Notification channel closed, resetting callbacks.')
+      firebase.messaging().onMessage(() => null)
+      firebase.messaging().onNotificationOpenedApp(() => null)
+    }
+
+    firebase.messaging().onMessage((message) => {
+      Logger.info(TAG, 'Notification received while open')
+      emitter({
+        message,
+        stateType: NotificationReceiveState.APP_ALREADY_OPEN,
+      })
+    })
+
+    firebase.messaging().onNotificationOpenedApp((message) => {
+      Logger.info(TAG, 'App opened via a notification')
+      emitter({
+        message,
+        stateType: NotificationReceiveState.APP_FOREGROUNDED,
+      })
+    })
+    return unsubscribe
+  })
+
+  Logger.debug(TAG, `setupMessage action: ${JSON.stringify(action)}`)
+
+  const isAppMounted = action.type === AppActions.APP_MOUNTED
+  if (isAppMounted) {
+    // TODO(erdal): is spawn the right one to use here?
+    yield spawn(watchFirebaseNotificationChannel, channelOnNotification)
+  } else {
+    channelOnNotification.close()
+  }
+
+  // Manual type checking because yield calls can't infer return type yet :'(
+  const initialNotification: Awaited<ReturnType<
+    FirebaseMessagingTypes.Module['getInitialNotification']
+  >> = yield call([firebase.messaging(), 'getInitialNotification'])
+  if (initialNotification) {
+    Logger.info(TAG, 'App opened fresh via a notification', JSON.stringify(initialNotification))
+    yield call(handleNotification, initialNotification, NotificationReceiveState.APP_OPENED_FRESH)
+  }
```

Comment:
Looks like this shouldn't happen when `AppActions.APP_UNMOUNTED` initiated this, right?

Reply:
Yep, good catch! moved it inside the if statement so we only check it on app mounted.